dom.bs

```
@@ -6519,6 +6529,11 @@ string <var>namespace</var> (default null):</p>
 
  <li><p>If <var>oldAttr</var> is <var>attr</var>, return <var>attr</var>.
 
+ <li><p>Let <var>verifiedValue</var> be the result of calling <a>verify attribute value</a>
+ <var>attr</var>'s <a for=Attr>value</a> for <var>attr</var>, with <var>element</var>.
+
+ <li><p>Set <var>attr</var>'s <a for=Attr>value</a> to <var>verifiedValue</var>.
+
```

Hmm, so doing the verification may run scripts. And that means oldAttr might not be anymore in the element it used to be.   Could that cause issues? Could the value be validated first for certain kind of element but then used on some other kind of element?

---

The default policy doesn't provide context about which element an attribute is set on only the name of the attribute.  In this case this algorithm is triggered by APIs such as setAttributeNode or setNamedItem.

So I don't think there's anything that can happen here that's too bad. Also any mechanism you use inside of the default policy will itself trigger the default policy so it should be fine?

---

The callback may have added the new attr to be in some other element's attribute list

---

FWIW, in https://bugzilla.mozilla.org/show_bug.cgi?id=1944511 I'm adding new tests and will try to make sure this case and the other ones mentioned below are covered.

But my guess would be we should just throw the same exception as in step 1 in that case:

> 1 If attr’s [element](https://dom.spec.whatwg.org/#concept-attribute-element) is neither null nor element, [throw](https://webidl.spec.whatwg.org/#dfn-throw) an "[InUseAttributeError](https://webidl.spec.whatwg.org/#inuseattributeerror)" [DOMException](https://webidl.spec.whatwg.org/#idl-DOMException).

---

The default policy only getting attribute information is an issue elsewhere, btw. https://github.com/w3c/trusted-types/issues/569 - Would be useful to add element information imho.

---

Good catch, I've added the fix suggested by Fred as that seems the most logical solution. I'll make sure to add a test against this when I fix it in WebKit.

---

Just for the record, this is https://phabricator.services.mozilla.com/D236107?id=977466#inline-1311431 and I agree with @smaug---- this is a problem in my current patch. I also started to add test for this and other cases in https://phabricator.services.mozilla.com/D236161#change-XQGcauHgW6CO

---

This is fixed and has a test, per Daniels' suggestion and my recent spec change this should be even cleaner. It will require a test change to verify the order of exceptions (the TT TypeError / default policy exception is now the first exception thrown)